wal/wal.go

```diff
@@ -899,6 +884,232 @@ func (r *Reader) Offset() int64 {
 	return r.total
 }
 
+// NewLiveReader returns a new live reader.
+func NewLiveReader(r io.Reader) *LiveReader {
+	return &LiveReader{rdr: r}
+}
+
+// Reader reads WAL records from an io.Reader. It buffers partial record data for
+// the next read.
+type LiveReader struct {
+	rdr        io.Reader
+	err        error
+	rec        []byte
+	hdr        [recordHeaderSize]byte
+	buf        [pageSize]byte
+	readIndex  int   // index in buf to start at for next read
+	writeIndex int   // index in buf to start at for next write
+	total      int64 // total bytes processed.
+	index      int   // used to track partial records
+}
+
+func (r *LiveReader) Err() error {
+	return r.err
+}
+
+func (r *LiveReader) TotalRead() int64 {
+	return r.total
+}
+
+func (r *LiveReader) fillBuffer() error {
+	n, err := r.rdr.Read(r.buf[r.writeIndex:len(r.buf)])
+	// We expect to get EOF, since we're reading the segment file as it's being written.
+	if err != nil && err != io.EOF {
+		return err
+	}
+	r.writeIndex += n
+	return nil
+}
+
+// Shift the buffer up to the read index.
+func (r *LiveReader) shiftBuffer() {
```

Comment:
maybe you can use ring buffer instead of `fillBuffer` and `shiftBuffer` to simplify it. 
https://www.youtube.com/watch?v=GbBrp6K7IvM

Reply:
I think that would just further complicate things. We would then need to make sure that every write doesn't overwrite data at the current `readIndex` and that reads don't require data past the current `writeIndex`.